### ISSUE TITLE
Device hang: untilize_with_unpadding on HEIGHT_SHARDED tensor with non-tile-aligned dims

### ISSUE THREAD

**Issue by rpavlovicTT** [tenstorrent/tt-metal#41504]
## Summary

`UntilizeWithUnpaddingDeviceOperation` on a HEIGHT_SHARDED tensor with non-tile-aligned dimensions deadlocks on wormhole_b0. All 64 worker cores enter a brisc↔ncrisc synchronization deadlock.

This is triggered by `ttnn.concat` on dim=3 when the concat dim has tile padding (`logical_shape[dim] != padded_shape[dim]`). The `build_untilize_rm_retilize_concat` massage path in `concat.cpp:120` calls:

```cpp
ttnn::untilize_with_unpadding(input_tensor, ttnn::Shape(ends), std::nullopt);
```

on each HEIGHT_SHARDED input. The `std::nullopt` memory config defaults to the input's HEIGHT_SHARDED config, and the resulting `UntilizeWithUnpaddingDeviceOperation` hangs.

## Minimal Reproducer

```python
import torch
import ttnn

device = ttnn.open_mesh_device(ttnn.MeshShape((1, 1)))

hs_config = ttnn.MemoryConfig(
    ttnn.TensorMemoryLayout.HEIGHT_SHARDED,
    ttnn.BufferType.L1,
    ttnn.ShardSpec(
        ttnn.CoreRangeSet(
            {ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(7, 7))}
        ),
        (512, 32),
        ttnn.ShardOrientation.ROW_MAJOR,
    ),
)

t1 = ttnn.from_torch(
    torch.randn(32, 32, 17, 16, dtype=torch.bfloat16),
    dtype=ttnn.bfloat16,
    layout=ttnn.TILE_LAYOUT,
    device=device,
    memory_config=hs_config,
)
t2 = ttnn.from_torch(
    torch.randn(32, 32, 17, 16, dtype=torch.bfloat16),
    dtype=ttnn.bfloat16,
    layout=ttnn.TILE_LAYOUT,
    device=device,
    memory_config=hs_config,
)
ttnn.synchronize_device(device)

# HANGS — internally calls untilize_with_unpadding on HEIGHT_SHARDED inputs
result = ttnn.concat([t1, t2], dim=3, memory_config=hs_config)
ttnn.synchronize_device(device)
```

Even simpler — `untilize_with_unpadding` directly:

```python
t = ttnn.from_torch(
    torch.randn(32, 32, 17, 16, dtype=torch.bfloat16),
    dtype=ttnn.bfloat16,
    layout=ttnn.TILE_LAYOUT,
    device=device,
    memory_config=hs_config,
)
ttnn.synchronize_device(device)

# HANGS
result = ttnn.untilize_with_unpadding(t, [31, 31, 16, 15])
ttnn.synchronize_device(device)
```

## tt-triage Output

```
dump_running_operations.py:
  Op 68075: UntilizeWithUnpaddingDeviceOperation
    logical_shape: [32, 32, 17, 16], dtype: BFLOAT16
    memory_layout: HEIGHT_SHARDED, buffer_type: L1
    shard_spec: ROUND_ROBIN_1D
  Prev Op 68074: InterleavedToShardedDeviceOperation
    logical_shape: [32, 32, 17, 16], dtype: BFLOAT16
    memory_layout: INTERLEAVED, buffer_type: L1

dump_callstacks.py:
  All 64 worker cores: brisc in wait_ncrisc_trisc(), ncrisc in wait_for_brisc_notification()
  cq_dispatch_subordinate stuck in wait_for_workers()
```

All health checks pass (ARC, NOC, ETH, CBs, binary integrity).

## Root Cause Analysis

The tensor `[32,32,17,16]` has **doubly non-tile-aligned dims**: dim2=17 and dim3=16 (tile size is 32x32). When `untilize_with_unpadding` operates on this tensor in HEIGHT_SHARDED layout (512x32 shards across 64 cores on 8x8 grid), the device kernel deadlocks.

The code path:
1. `ttnn::concat` on dim=3 with HEIGHT_SHARDED TILE tensors
2. `build_untilize_rm_retilize_concat` predicate fires: `logical_shape[3]=16 != padded_shape[3]=32`
3. Pre-transform calls `untilize_with_unpadding(input, ends=[31,31,16,15], nullopt)` — `concat.cpp:120`
4. `UntilizeWithUnpaddingDeviceOperation` on HEIGHT_SHARDED → **deadlock**

## Impact

Blocks phi-1 model execution in tt-xla (both `pytest` and `ttrt run` of the compiled flatbuffer hang).

## Environment

- Device: wormhole_b0, n150 board, ARC FW 18.12.1
- tt-metal commit: `aa301f48dcd8e9f1b9057a21a79ffc981f551d9f`
- tt-mlir commit: `c3d6b7f14e7d2cff04908c5058e62e20f498c376`

## Related Issues

- #33858 — `ttnn.slice` hangs when input is HEIGHT_SHARDED
- #39947 — Terminal hang after `ttnn.slice` execution
- #19475 — `UntilizeWithUnpadding` + HEIGHT_SHARDED + non-tile-aligned dims → PCC failure

**Comment by rpavlovicTT** [comment]
Fixed by #41521
